### ISSUE TITLE
Sending notification on behalf or another process

### ISSUE THREAD

**Issue by PradeepKiruvale** [lnicola/sd-notify#7]
Hi,

I am using this crate for sending the `notifications` to `systemd` on behalf of another process. 
Below is my code snippet that is used for sending the notification. I collect the `process ID` of the process to be monitored by the   `watchdog` program and send the notification to the systemd on behalf of that process. The code runs fine, without any errors. But the notifications are not reaching the system. So, the `systemd` keeps restarting the service after `watchdog timeout`. Please help me am I doing the right thing here or not.

fn notify_systemd(pid: u32) {

    let result = sd_notify::notify(true, &[NotifyState::MainPid(pid)]);

    dbg!(&result);

    match result {

        Ok(()) => {

            println!("notify pid: {}", pid);
        }

        Err(e) => {
            println!("error: {:?}", e.to_string());
        }

    }

Thanks,
Pradeep

**Comment by lnicola** [comment]
Hello! You're using `Type=notify`, right?

I don't know what might be causing this, but it's an interesting problem since it involves the interaction between two different features. How are you enabling the watchdog, using `WatchdogSec` in the service file, or using the `Watchdog` notification?

**Comment by PradeepKiruvale** [comment]
I was not using `Type=notify` because the service start was hanging. I am using `WatchdogSec` to enable the watchdog feature. Also, I noticed from the [documentation  ](https://docs.rs/sd-notify/0.4.0/sd_notify/fn.notify.html#:~:text=The%20implementation%20of%20this%20function%20is%20somewhat%20naive:%20it%20doesn%E2%80%99t%20support%20sending%20notifications%20on%20behalf%20of%20other%20processes%2C%20doesn%E2%80%99t%20pass%20file%20descriptors%2C%20doesn%E2%80%99t%20send%20credentials%2C%20and%20does%20not%20increase%20the%20send%20buffer%20size.%20It%E2%80%99s%20still%20useful%2C%20though%2C%20in%20usual%20situations.)that `sd_notify` crate does not support sending notifications on behalf of another process. So, I feel it does not work for my use case.

**Comment by lnicola** [comment]
> does not support sending notifications on behalf of another process

I wrote that a couple of years ago, I wonder what I meant :sweat_smile:. But `MAINPID` is not a notification on behalf of another process. It's a notification telling `systemd` to watch a different process instead of the current one.

**Comment by PradeepKiruvale** [comment]
> > does not support sending notifications on behalf of another process
> 
> I wrote that a couple of years ago, I wonder what I meant 😅. But `MAINPID` is not a notification on behalf of another process. It's a notification telling `systemd` to watch a different process instead of the current one.

Ok, that means I can use these API's to send the notification on behalf of another process. Good to know that. But as I understand the `sd-notification` feature, the `systemd` creates a socket to communicate between the service it starts and systemd. Will it be possible to use the same socket endpoint to send the message on behalf of another service?. I read that its possible, provided `if there are enough privileges`. But how to set these privileges in the service file to allow another process to send the notification.
> 
> And if you don't use `Type=notify`, I don't know if notifications do anything..

Ok, I did not use because there is no way to tell the systemd that the `ready` state of a process using PID. So, the systemd hangs there till it receives the notification telling it's ready.

**Comment by lnicola** [comment]
> Ok, that means I can use these API's to send the notification on behalf of another process. Good to know that.

You can't, because the docs say  it's not supported. I was just saying that `MAINPID` is sent on behalf of the current, not another process.

**Comment by PradeepKiruvale** [comment]
> > Ok, that means I can use these API's to send the notification on behalf of another process. Good to know that.
> 
> You can't, because the docs say it's not supported. I was just saying that `MAINPID` is sent on behalf of the current, not another process.

Ha ok, thanks for the clarification.

**Comment by lnicola** [comment]
Yeah, so see [`sd_pid_notify`](https://www.man7.org/linux/man-pages/man3/sd_notify.3.html), that's what I meant there.

**Comment by PradeepKiruvale** [comment]
> Yeah, so see [`sd_pid_notify`](https://www.man7.org/linux/man-pages/man3/sd_notify.3.html), that's what I meant there.

Yup, had a look. Needed some Rust crate that supports this. I mean pure Rust crate :).

**Comment by lnicola** [comment]
Well, we can implement it, but I wasn't sure about how the API should look.

**Comment by PradeepKiruvale** [comment]
I feel to be in line with `C` APIs we can implement `notify` and `notify_pid` as two separate APIs.

**Comment by lnicola** [comment]
So I just have to figure out how to send the PID...

**Comment by maowen** [comment]
One thing I learned is that I needed to set `NotifyAccess=all` so my child process was able to notify over the socket.
